chain/store/splitstore/splitstore.go

```diff
@@ -0,0 +1,414 @@
+		return nil, err
+	}
+
+	chCold, err := s.cold.AllKeysChan(ctx)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+
+	ch := make(chan cid.Cid)
+	go func() {
+		defer cancel()
+		defer close(ch)
+
+		for _, in := range []<-chan cid.Cid{chHot, chCold} {
+			for cid := range in {
+				select {
+				case ch <- cid:
+				case <-ctx.Done():
+					return
+				}
+			}
+		}
+	}()
+
+	return ch, nil
+}
+
+func (s *SplitStore) HashOnRead(enabled bool) {
+	s.hot.HashOnRead(enabled)
+	s.cold.HashOnRead(enabled)
+}
+
+func (s *SplitStore) View(cid cid.Cid, cb func([]byte) error) error {
+	err := s.hot.View(cid, cb)
+	switch err {
+	case bstore.ErrNotFound:
+		return s.cold.View(cid, cb)
+
+	default:
+		return err
+	}
+}
+
+// State tracking
+func (s *SplitStore) Start(cs *store.ChainStore) error {
+	s.cs = cs
+	s.curTs = cs.GetHeaviestTipSet()
+
+	// load base epoch from metadata ds
+	// if none, then use current epoch because it's a fresh start
+	bs, err := s.ds.Get(baseEpochKey)
+	switch err {
+	case nil:
+		s.baseEpoch = bytesToEpoch(bs)
+
+	case dstore.ErrNotFound:
+		err = s.setBaseEpoch(s.curTs.Height())
+		if err != nil {
+			return err
+		}
+
+	default:
+		return err
+	}
+
+	// watch the chain
+	cs.SubscribeHeadChanges(s.HeadChange)
+
+	return nil
+}
+
+func (s *SplitStore) Close() error {
+	if s.isCompacting() {
+		log.Warn("ongoing compaction; waiting for it to finish...")
+		for s.isCompacting() {
+			time.Sleep(time.Second)
+		}
+	}
+
+	return s.env.Close()
+}
+
+func (s *SplitStore) HeadChange(revert, apply []*types.TipSet) error {
+	s.curTs = apply[len(apply)-1]
+	epoch := s.curTs.Height()
+	if epoch-s.baseEpoch > CompactionThreshold && !s.isCompacting() {
+		s.setCompacting(true)
+		go func() {
+			defer s.setCompacting(false)
+
+			log.Info("compacting splitstore")
+			start := time.Now()
+
+			s.compact()
+
+			log.Infow("compaction done", "took", time.Since(start))
+		}()
+	}
+
+	return nil
+}
+
+func (s *SplitStore) isCompacting() bool {
+	s.stateMx.Lock()
+	defer s.stateMx.Unlock()
+	return s.compacting
+}
+
+func (s *SplitStore) setCompacting(state bool) {
+	s.stateMx.Lock()
+	defer s.stateMx.Unlock()
+	s.compacting = state
+}
+
+// Compaction/GC Algorithm
+func (s *SplitStore) compact() {
+	// create two on disk live sets, one for marking the cold finality region
+	// and one for marking the hot region
+	hotSet, err := NewLiveSet(s.env, "hot")
+	if err != nil {
+		// TODO do something better here
+		panic(err)
+	}
+	defer hotSet.Close() //nolint:errcheck
+
+	coldSet, err := NewLiveSet(s.env, "cold")
+	if err != nil {
+		// TODO do something better here
+		panic(err)
+	}
+	defer coldSet.Close() //nolint:errcheck
+
+	// Phase 1: marking
+	log.Info("marking live objects")
+	startMark := time.Now()
+
+	// Phase 1a: mark all reachable CIDs in the hot range
+	curTs := s.curTs
+	epoch := curTs.Height()
+	coldEpoch := s.baseEpoch + build.Finality
+	err = s.cs.WalkSnapshot(context.Background(), curTs, epoch-coldEpoch+1, false, false,
```

Comment:
Definitely need to address #4994 given that we're relying on WalkSnapshot in production code now!

Reply:
Doesn't look thaaat critical, it's just less performant than it could be.

Reply:
It's O(n^2) and therefore highly sensitive to the size of the reacheable nodes. Not an algorithm we want in production, especially when we're trying to curtail allows.

Reply:
It's not O(n^2), just not as efficient as it should be.
Optimizing it is out of scope for this PR I think, we can do it separately.

Reply:
I updated the description of this PR. It wouldn't hurt to explain what's being contributed here -- there are a lot more people watching the repo than people that were privy to this design discussion.